src/types.md

```diff
@@ -365,30 +365,101 @@ x = bo(5,7);
-  : The closure can be called once. A closure called as `FnOnce` can move out
-    of its captured values.
+impl<'a> (FnOnce() -> String) for Closure<'a> {
+    fn call_once(self) -> String {
+        self.s += &*self.t;
+        self.s
+    }
+}
+```
 
-* `FnMut`
-  : The closure can be called multiple times as mutable. A closure called as
-    `FnMut` can mutate values from its environment. `FnMut` inherits from
-    `FnOnce` (i.e. anything implementing `FnMut` also implements `FnOnce`).
+so that the call to `f` works as if it were:
 
-* `Fn` : The closure can be called multiple times through a shared reference. A
-    closure called as `Fn` can neither move out from nor mutate captured
-    variables, but read-only access to such values is allowed. Using `move` to
-    capture variables by value is allowed so long as they aren't mutated or
-    moved in the body of the closure. `Fn` inherits from `FnMut`, which itself
-    inherits from `FnOnce`.
+```rust,ignore
+f(Closure{s: s, t: &t});
+```
 
-Closures that don't use anything from their environment, called *non-capturing
-closures*, can be coerced to function pointers (`fn`) with the matching
-signature. To adopt the example from the section above:
+The compiler prefers to capture a closed-over variable by immutable borrow,
+followed by mutable borrow and finally by move (or copy, for [`Copy`] types). It
+will pick the first choice of these that allows the closure to compile. If the
+`move` keyword is used, then all captures are by move or copy, regardless of
+whether a borrow would work. The `move` keyword is usually used to allow the
+closure to outlive the captured values, such as if the closure is being returned
+or used to spawn a new thread.
+
+Structs and tuples are always captured entirely, not by individual fields. It
+may be necessary to borrow into a local variable in order to capture a single
+field:
+
+```rust
+# use std::collections::HashSet;
+# 
+struct SetVec {
+    set: HashSet<u32>,
+    vec: Vec<u32>
+}
+
+impl SetVec {
+    fn populate(&mut self) {
+        let vec = &mut self.vec;
+        self.set.iter().for_each(|&n| {
+            vec.push(n);
+        })
+    }
+}
+```
+
+If, instead, the closure were to use `self.vec` directly, then it would attempt
+to capture `self` by mutable reference. But since `self.set` is already
+borrowed to iterate over, the closure would not compile.
+
+### Call traits and coercions
+
+Closure types all implement `[FnOnce]`, indicating that they can be called once
+by consuming ownership of the closure. Additionally, some closures implement
+more specific call traits:
+
+* A closure which does not move out of any captured variables implements
+  `[FnMut]`, indicating that it can be called by mutable reference.
+
+* A closure which does not mutate or move out of any captured variables
+  implements `[Fn]`, indicating that it can be called by shared reference.
+
+> Note that `move` closures may still implement `[Fn]` or `[FnMut]`, even
```

Comment:
Note needs to have a colon after it to remain consistent with the rest of the notes in the reference.